src/main/java/io/stargate/sgv2/jsonapi/service/operation/reranking/FeatureUsage.java

```diff
@@ -0,0 +1,125 @@
+package io.stargate.sgv2.jsonapi.service.operation.reranking;
+
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a collection of {@link Feature}s in a command. This class is immutable and can be used
+ * to track which features are utilized in a command. This class is immutable; methods that modify
+ * the set of features return a new {@code FeatureUsage} instance. It uses an {@link EnumSet}
+ * internally for efficient storage and operations on features.
+ */
+public final class FeatureUsage {
+  private final EnumSet<Feature> features;
+
+  /** A instance representing no features in use. */
+  public static final FeatureUsage EMPTY = new FeatureUsage(EnumSet.noneOf(Feature.class));
+
+  private FeatureUsage(EnumSet<Feature> features) {
+    this.features = EnumSet.copyOf(features);
+  }
+
+  /**
+   * Creates a {@code FeatureUsage} instance from an array of {@link Feature}s.
+   *
+   * @param features The features to include. If null or empty, {@link #EMPTY} is returned.
+   * @return A new {@code FeatureUsage} instance containing the specified features.
+   */
+  public static FeatureUsage of(Feature... features) {
+    if (features == null || features.length == 0) {
+      return EMPTY;
+    }
+    EnumSet<Feature> set = EnumSet.noneOf(Feature.class);
+    Collections.addAll(set, features);
+    return new FeatureUsage(set);
+  }
+
+  /**
+   * Creates a {@code FeatureUsage} instance from an {@link EnumSet} of {@link Feature}s.
+   *
+   * @param features The set of features to include. If null or empty, {@link #EMPTY} is returned.
+   * @return A new {@code FeatureUsage} instance containing the specified features.
+   */
+  public static FeatureUsage of(EnumSet<Feature> features) {
+    if (features == null || features.isEmpty()) {
+      return EMPTY;
+    }
+    return new FeatureUsage(features);
+  }
+
+  /**
+   * Returns a new {@code FeatureUsage} instance that includes the specified feature in addition to
+   * the features in this instance. If the feature is already present, this instance is returned.
+   *
+   * @param feature The feature to add. Must not be null.
+   * @return A new {@code FeatureUsage} instance with the added feature, or this instance if the
+   *     feature was already present.
+   * @throws NullPointerException if the feature is null.
+   */
+  public FeatureUsage withFeature(Feature feature) {
+    Objects.requireNonNull(feature, "Feature cannot be null");
+    if (this.features.contains(feature)) {
+      return this;
+    }
+    EnumSet<Feature> newSet = EnumSet.copyOf(this.features);
+    newSet.add(feature);
+    return new FeatureUsage(newSet);
```

Comment:
Wrt copying: this `EnumSet` does not need to be copied in constructor as it is not referenced by anything else.

Reply:
+1 there is a double copy happening here